leakpro/attacks/minv_attacks/plgmi.py

```diff
@@ -0,0 +1,200 @@
+        self.dis_beta2 = configs.get("dis_beta2", 0.9)
+        # Paths
+        self.pseudo_label_path = configs.get("pseudo_label_path")
+        self.output_dir = configs.get("output_dir")
+        # Define the validation dictionary as: {parameter_name: (parameter, min_value, max_value)}
+        validation_dict = {
+            # alpha, 0 to inf
+            "alpha": (self.alpha, 0, 1000), # 0 to inf
+            "n_dis": (self.n_dis, 1, 1000), # 1 to inf
+            # TODO: Add more parameters
+        }
+
+        # Validate parameters
+        for param_name, (param_value, min_val, max_val) in validation_dict.items():
+            self._validate_config(param_name, param_value, min_val, max_val)
+
+
+    def description(self:Self) -> dict:
+        """Return the description of the attack."""
+        title_str = "PLG-MI Attack"
+        reference_str = "Pseudo Label-Guided Model Inversion Attack via Conditional Generative \
+                            Adversarial Network, Yuan et al. 2023, https://arxiv.org/abs/2302.09814"
+        summary_str = "This attack is a model inversion attack that uses the PLG-MI algorithm."
+        detailed_str = "The Pseudo Label Guided Model Inversion Attack (PLG-MI) is a white-box attack \
+                        that implements pseudo-labels on a public dataset to construct a conditional GAN. \
+                            Steps: \
+                                1. Top-n selection of pseudo labels. \
+                                2. Train the GAN. \
+                                3. Generate samples from the GAN. \
+                                4. Compute image metrics. "
+        return {
+            "title_str": title_str,
+            "reference": reference_str,
+            "summary": summary_str,
+            "detailed": detailed_str,
+        }
+
+
+    def top_n_selection(self:Self) -> DataLoader:
+        """"Top n selection of pseudo labels."""
+        # TODO: This does not scale well. Consider creating a class for the dataloader and implementing the __getitem__ method.
+        logger.info("Performing top-n selection for pseudo labels")
+        self.target_model = self.handler.target_model
+        self.target_model.eval()
+        all_confidences = []
+        for images, _ in self.public_dataloader:
+            with torch.no_grad():
+                outputs = self.target_model(images)
+                confidences = F.softmax(outputs, dim=1)
+                all_confidences.append(confidences)
+        # Concatenate all confidences
+        self.confidences = torch.cat(all_confidences)
+
+        logger.info("Retrieved confidences from the target model")
+        # Get the pseudo labels
+        pseudo_labels = torch.max(self.confidences, dim=1)
+
+        # Empty array of size num_classes to store the pseudo labels
+        pseudo_map = [[] for _ in range(self.num_classes)]
+
+        for i, (conf, label) in enumerate(zip(pseudo_labels[0], pseudo_labels[1])):
```

Comment:
Is this correct? Shouldnt pseudo_labels just contain the label for each image? Are conf in it too? NOTE: I havent run this yet as the data takes forever to get in place.